File: internal/sheet/sheet.go
// Package sheet holds one in-memory table per workspace. It imports nothing
// from the packages above it, which is what lets it be tested with no display
// attached (I-6).
package sheet

import "github.com/flacyak/uno/internal/formula"

// Sheet is the one copy of the data in memory. ingest builds it at load, the
// grid reads it every frame, and Set is the single door through which it
// changes (I-7: only ever from the UI goroutine).
type Sheet struct {
	Name string
	// Source describes how ingest read these bytes, for the status bar to show
	// verbatim. It is an opaque label here on purpose: ingest owns the wording,
	// so sheet still knows nothing about delimiters or encodings (I-5).
	Source  string
	Columns []Column
	rows    [][]string
	edits   []Edit // every change made to rows, in order; see edit.go

	// computed is what Display hands out where something has filled it in: a
	// column a formula is bound to, and a cell holding notation. It is indexed
	// by column and nil for a column nothing computes, so the common case costs
	// one bounds check and no allocation on the per-frame path (I-1).
	//
	// A slice and not a map because Display runs about two hundred times a
	// frame, and hashing a key that many times to answer "no" for most of them
	// is work the grid does not have to do.
	computed [][]string

	// bound is the expression each computed column resolves to, and graph is
	// what they depend on. Both are rebuilt by replaying the log, so neither is
	// written to the file: raw plus the log is the whole truth of a workspace
	// (I-4), and a second copy of a binding is a second thing to keep in step.
	bound map[int]formula.Formula
	graph formula.Graph
}

// Column pairs a header with the kind inferred from the values beneath it.
type Column struct {
	Header string
	Kind   Kind // inferred at load; never stored in the file
	// Flagged marks a column that looks numeric but does not parse cleanly,
	// such as one whose thousands separators break half its values. It is the
	// condition M2's pattern recogniser acts on.
	Flagged bool
}

// New builds a sheet from a header row and the data rows beneath it, inferring
// each column's kind as it goes. A row longer than the header contributes no
// column: the header decides the shape, and At tolerates the overhang.
func New(name string, header []string, rows [][]string) *Sheet {
	s := &Sheet{
		Name:     name,
		Columns:  make([]Column, len(header)),
		rows:     rows,
		computed: make([][]string, len(header)),
	}
	for i, h := range header {
		s.Columns[i].Header = h
		s.inferKindOf(i)
	}
	return s
}

func (s *Sheet) Rows() int { return len(s.rows) }
func (s *Sheet) Cols() int { return len(s.Columns) }

// Raw is what the cell stores. The log records it, undo restores it, and the
// recogniser reads it, because all three are about the value a person put there
// rather than the one they are being shown.
//
// An absent cell is empty rather than an error: short rows are normal in real
// exports, and returning "" is what lets the table skip bounds checks while
// scrolling.
func (s *Sheet) Raw(row, col int) string {
	if row < 0 || row >= len(s.rows) {
		return ""
	}
	if col < 0 || col >= len(s.rows[row]) {
		return ""
	}
	return s.rows[row][col]
}

// Display is what the cell shows, and what the grid binds to. It is called once
// per visible cell on every scroll frame (I-1), so it reads and returns:
// whatever fills it in does so when an edit lands, never when a cell is read.
//
// The body is a cache read and nothing else. Recalculation fills it when an
// edit lands, walking the dependency graph in order; a notation cell fills its
// own entry when it is authored. Evaluating here instead would put an expression
// tree walk on the scroll path, which is the exact mistake this seam exists to
// prevent.
func (s *Sheet) Display(row, col int) string {
	if col >= 0 && col < len(s.computed) {
		if vals := s.computed[col]; row >= 0 && row < len(vals) {
			if v := vals[row]; v != "" {
				return v
			}
		}
	}
	return s.Raw(row, col)
}

File: internal/grid/src/sheet/sheet.ts
// Package sheet holds one table per workspace. It imports nothing from the
// packages above it, which is what lets it be tested with no display attached.
//
// A Sheet is a table held in memory: its source rows, a Schema folded from the
// log, and the rows finished from both as they are read. The engine does the
// same over a file it never holds, with the same Schema and the same pipeline,
// so the rules a test pins here are the rules a 30 GB file follows.

import type { Formula } from "../formula/index.ts";
import { evaluate } from "../formula/index.ts";
import type { Program } from "../program/index.ts";
import { text as programText } from "../program/index.ts";
import type { Edit } from "./edit.ts";
import { NO_ROW, Op } from "./edit.ts";
import type { Kind } from "./kind.ts";
import { inferKind } from "./kind.ts";
import type { Finished } from "./pipeline.ts";
import { finishRows, formatValue } from "./pipeline.ts";
import { Schema } from "./schema.ts";
import type { Written } from "./schema.ts";

/** Column pairs a header with the kind inferred from the values beneath it. */
export interface Column {
  header: string;
  /** Inferred at load; never stored in the file. */
  kind: Kind;
  /**
   * Marks a column that looks numeric but does not parse cleanly, such as one
   * whose thousands separators break half its values. It is the condition the
   * recogniser acts on.
   */
  flagged: boolean;
}

/**
 * FINISH_ROWS is how many rows are finished together when one of them is read:
 * the engine's block size, so a sheet and a file compute formulas in the same
 * steps.
 */
const FINISH_ROWS = 1024;

export class Sheet {
  readonly columns: Column[];
  /** How `ingest` read these bytes, for the status bar to show verbatim. It is
   * an opaque label here on purpose: `ingest` owns the wording, so `sheet` still
   * knows nothing about delimiters or encodings. */
  source = "";

  private readonly rowData: readonly (readonly string[])[];
  private readonly schema: Schema;

  /**
   * Rows finished since the log last changed, indexed by row.
   *
   * `display` runs about two hundred times a frame, so it has to be an array
   * read. The first read of a row after an edit finishes it; every read after
   * that is the lookup.
   */
  private finished: Array<Finished | undefined> = [];

  /**
   * Builds a sheet from a header row and the data rows beneath it, inferring
   * each column's kind as it goes.
   *
   * A row longer than the header contributes no column: the header decides the
   * shape.
   */
  constructor(
    readonly name: string,
    header: string[],
    rows: string[][],
  ) {
    this.rowData = rows;
    this.schema = new Schema(header, rows.length);
    this.columns = header.map((h) => ({ header: h, kind: "text" as Kind, flagged: false }));
    this.infer();
  }

  rows(): number {
    return this.rowData.length;
  }

  cols(): number {
    return this.columns.length;
  }

  /**
   * raw is what the cell stores. The log records it, undo restores it, and the
   * recogniser reads it, because all three are about the value a person put
   * there rather than the one they are being shown.
   *
   * An absent cell is empty rather than an error: short rows are normal in real
   * exports, and returning "" is what lets the table skip bounds checks while
   * scrolling.
   */
  raw(row: number, col: number): string {
    if (col < 0) return "";
    return this.row(row)?.raw[col] ?? "";
  }

  /**
   * display is what the cell shows, and what the grid binds to: raw, except
   * where notation or a formula fills the cell in.
   */
  display(row: number, col: number): string {
    if (col < 0) return "";
    return this.row(row)?.shown[col] ?? "";
  }

  /** written is the last write into a cell, or undefined where nobody typed. */
  written(row: number, col: number): Written | undefined {
    return this.schema.writtenIn(row)?.get(col);
  }

  /**
   * row finishes the block a row is in, not the row alone, so that a bound
   * column is computed over the block a column at a time.
   */
  private row(row: number): Finished | undefined {
    if (row < 0 || row >= this.rowData.length) return undefined;
    const f = this.finished[row];
    if (f !== undefined) return f;

    const first = row - (row % FINISH_ROWS);
    const block = finishRows(this.schema, first, this.rowData.slice(first, first + FINISH_ROWS));
    for (let i = 0; i < block.length; i++) this.finished[first + i] = block[i];
    return block[row - first];
  }

  // ------------------------------------------------------------ the log

  /**
   * set records a value typed into one cell. It is the only way the grid
   * changes a single value, so the log can never fall behind the data it
   * describes.
   */
  set(row: number, col: number, v: string): void {
    this.record({ seq: 0, op: Op.Set, row, col, was: this.raw(row, col), now: v });
  }

  /**
   * apply runs a program over every value in a column and records it as one
   * operation.
   *
   * It is the door a pattern proposal comes through, and the reason the log
   * stays proportional to what a person did rather than to how much data they
   * did it to: 3,149 cells change and one line is written.
   */
  apply(col: number, p: Program): void {
    this.record({ seq: 0, op: Op.Apply, row: NO_ROW, col, now: programText(p) });
  }

  /**
   * note puts notation in one cell: the person types markdown and the cell
   * shows the symbols it describes.
   *
   * A notation cell stores its source and shows what the source describes,
   * which is the same relationship a bound column has to its expression. It
   * reads no columns and joins no dependency graph, which is everything else.
   */
  note(row: number, col: number, src: string): void {
    this.record({ seq: 0, op: Op.Note, row, col, was: this.raw(row, col), now: src });
  }

  /**
   * bind makes a column derived: what it shows is computed from the columns
   * the expression names. It records one line for a column of any length, and
   * the expression is what the file carries, not the results.
   */
  bind(col: number, f: Formula): void {
    this.record({ seq: 0, op: Op.Bind, row: NO_ROW, col, now: f.toString() });
  }

  /**
   * unbind takes the formula off a column. What it stored before it was bound
   * is what it shows again: binding never removed those values.
   *
   * It is recorded, so it can be undone, and it refuses a column nothing is
   * bound to rather than doing nothing quietly.
   */
  unbind(col: number): void {
    if (col < 0 || col >= this.columns.length) {
      throw new Error(`column ${col} is outside the ${this.columns.length} columns of this sheet`);
    }
    const was = this.binding(col);
    if (was === undefined) {
      throw new Error(`no formula is bound to ${this.columns[col]!.header}`);
    }
    this.record({ seq: 0, op: Op.Unbind, row: NO_ROW, col, was, now: "" });
  }

  /** binding reports the expression a column resolves to. */
  binding(col: number): string | undefined {
    return this.schema.binding(col);
  }

  private record(e: Edit): void {
    this.schema.record(e);
    this.finished = [];
    this.infer();
  }

  /**
   * replay applies a log and keeps it, so that saving a reopened file preserves
   * the history rather than starting a new one. It is linear in the number of
   * operations, not in the rows they touch.
   */
  replay(edits: Edit[]): void {
    try {
      this.schema.replay(edits);
    } finally {
      this.finished = [];
    }
    this.infer();
  }

  /**
   * logEquals reports whether this sheet's log is exactly the one given. It
   * compares rather than counting, because undo makes a count ambiguous.
   */
  logEquals(other: Edit[]): boolean {
    return this.schema.logEquals(other);
  }

  editCount(): number {
    return this.schema.editCount();
  }

  /** edits returns the log to be written, copied. */
  edits(): Edit[] {
    return this.schema.edits();
  }

  // ------------------------------------------------------------ naming

  resolve(name: string): number {
    return this.schema.resolve(name);
  }

  /**
   * evaluateAt computes one row of an expression without binding it, which is
   * what the editor's preview is: an answer to "what would this do", read the
   * way a binding reads the row.
   */
  evaluateAt(f: Formula, row: number): string {
    const shown = this.row(row)?.shown ?? [];
    return formatValue(
      evaluate(f, {
        value: (name) => {
          const i = this.schema.indexOf(name);
          return i === undefined ? undefined : (shown[i] ?? "");
        },
      }),
    );
  }

  /**
   * infer re-reads every column's sample and renames it. A column whose last
   * unparseable value was just fixed is a number column now, and a column that
   * reads it may have become one too.
   */
  private infer(): void {
    for (let col = 0; col < this.columns.length; col++) {
      const { kind, flagged } = inferKind(this.rowData.length, (row) => this.display(row, col));
      const c = this.columns[col]!;
      c.kind = kind;
      c.flagged = flagged;
    }
  }
}
